crates/red_knot_workspace/src/project/combine.rs

```diff
@@ -0,0 +1,155 @@
+use std::{collections::HashMap, hash::BuildHasher};
+
+use crate::project::options::{EnvironmentOptions, SrcOptions};
+use red_knot_python_semantic::{PythonPath, PythonPlatform, PythonVersion};
+use ruff_db::system::SystemPathBuf;
+
+/// Combine two values, preferring the values in `self`.
+///
+/// The logic should follow that of Cargo's `config.toml`:
+///
+/// > If a key is specified in multiple config files, the values will get merged together.
+/// > Numbers, strings, and booleans will use the value in the deeper config directory taking
+/// > precedence over ancestor directories, where the home directory is the lowest priority.
+/// > Arrays will be joined together with higher precedence items being placed later in the
+/// > merged array.
+///
+/// Note: The merging behavior differs from uv in that values with higher precedence in arrays
+/// are placed later in the merged array. This is because we want to support overriding previously
+/// provided values, including unsetting them. For example: patterns coming last in file inclusion and exclusion patterns
+/// allow overriding earlier patterns, matching the `gitignore` behavior. This also closer matches
+/// a CLI interface where values provided later can override values provided earlier: `knot --ignore=correctness --warn=correctness/rule`
+/// should enable `correctness/rule` even though `correctness` was previously ignored.
```

Comment:
This particular example is a bit confusing to me, as `--ignore` and `--warn` seem like two (related but) different options?

The inclusion/exclusion patterns might be a better example a merging of something like `exclude = ["*.png"]` and `exclude = ["!important.png"]` to `["*.png", "!important.png"]`?

Reply:
I updated the comment and also pointed out a potential "shortcoming" of this approach. I'm interested to hear if there are any concerns or if we feel comfortable moving forward with how it is

Reply:
I can certainly imagine cases where we would want a "replace" logic for lists, instead of a "extend" logic. But I think it's completely fine to try and see how far this approach takes us. Thank you for the updated description!